querybook/server/lib/query_analysis/templating.py

```diff
@@ -26,179 +30,241 @@ class QueryJinjaSyntaxException(QueryTemplatingError):
-        comment_re, lambda match: "{{ " + json.dumps(match.group()) + " }}", query
-    )
+class TemplatedQueryRenderer(object):
+    def __init__(self, engine_id: int = None):
+        self.env = SandboxedEnvironment()
+        self.engine_id = engine_id
+        self.env.globals.update(latest_partition=self.get_latest_partition)
 
+    @staticmethod
+    def _escape_sql_comments(query: str):
+        return re.sub(
+            comment_re, lambda match: "{{ " + json.dumps(match.group()) + " }}", query
+        )
 
-def _detect_cycle_helper(node: str, dag: _DAG, seen: Set[str]) -> bool:
-    if node in seen:
-        return True
-
-    seen.add(node)
-
-    children = dag.get(node, [])
-    for child in children:
-        if _detect_cycle_helper(child, dag, seen):
+    @staticmethod
+    def _detect_cycle_helper(node: str, dag: _DAG, seen: Set[str]) -> bool:
+        if node in seen:
             return True
-    seen.remove(node)
-    return False
-
-
-def _detect_cycle(dag: _DAG) -> bool:
-    seen = set()
-    return any(_detect_cycle_helper(node, dag, seen) for node in dag.keys())
-
-
-def get_default_variables():
-    return {
-        "today": datetime.today().strftime("%Y-%m-%d"),
-        "yesterday": (datetime.today() - timedelta(1)).strftime("%Y-%m-%d"),
-    }
-
-
-def get_templated_variables_in_string(s: str) -> Set[str]:
-    """Find possible templated variables within a string
-
-    Arguments:
-        s {str}
-    Returns:
-        Set[str] - set of variable names
-    """
-
-    env = SandboxedEnvironment()
-    ast = env.parse(s)
-    variables = meta.find_undeclared_variables(ast)
-
-    # temporarily applying https://github.com/pallets/jinja/pull/994/files
-    # since the current version is binded by flask
-    filtered_variables = set()
-    for variable in variables:
-        if variable not in env.globals:
-            filtered_variables.add(variable)
-
-    return filtered_variables
 
+        seen.add(node)
+
+        children = dag.get(node, [])
+        for child in children:
+            if TemplatedQueryRenderer._detect_cycle_helper(child, dag, seen):
+                return True
+        seen.remove(node)
+        return False
+
+    @staticmethod
+    def _detect_cycle(dag: _DAG) -> bool:
+        seen = set()
+        return any(
+            TemplatedQueryRenderer._detect_cycle_helper(node, dag, seen)
+            for node in dag.keys()
+        )
 
-def verify_all_variables_are_defined(variables_required, variables_provided):
-    for variable_name in variables_required:
-        if variable_name not in variables_provided:
-            raise UndefinedVariableException(
-                "Invalid variable name {}".format(variable_name)
+    @staticmethod
+    def get_default_variables():
+        return {
+            "today": datetime.today().strftime("%Y-%m-%d"),
+            "yesterday": (datetime.today() - timedelta(1)).strftime("%Y-%m-%d"),
+        }
+
+    @with_session
+    def get_latest_partition(
```

Comment:
you can make this curried by generating a function based on self.engine_id, for example
```
def create_get_latest_partition(self, engine_id):
    metastore_loader = None
    with DBSession() as session:
        engine = ...
        metastore_id = ...
       metastore_loader = ...
    
    def get_latest_partition():
        ....
    
    return get_latest_partition